app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,160 @@
+        if (previousPosition != null) {
+            // but moved -> remove
+            if (previousPosition != position) remove(key)
+            else return true
+        }
+        val tileInCache = byTile[getTilePosFor(position)]
+        if (tileInCache  != null) {
+            tileInCache.add(key)
+            byKey[key] = position
+            return true
+        }
+        return false
+    }
+
+    // returns keys that were not put to cache because they are not in fully contained tiles
+    // contrary to get(bbox), here the bbox may be larger than what can be put in cache
+    fun replaceAllInBBox(entries: Collection<Pair<T, LatLon>>, bbox: BoundingBox): List<T> {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val completelyContainedTiles = tiles.filter { it.asBoundingBox(tileZoom).isCompletelyInside(bbox) }
+        if (tiles != completelyContainedTiles)
+            Log.i(TAG, "replacing tiles not completely contained in BBox")
+        val tilesToFetch = if (completelyContainedTiles.size > maxTiles) {
+            Log.i(TAG, "trying to replacing more tiles than fit in cache, ignoring some tiles")
+            completelyContainedTiles.subList(0,maxTiles-1)
+        } else
+            completelyContainedTiles
+        val ignoredKeys = replaceAllInTiles(entries, tilesToFetch)
+
+        // trim only when bbox!
+        trim()
+        return ignoredKeys
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(entries: Collection<Pair<T, LatLon>>, tiles: Collection<TilePos>): List<T> {
+        // create / replace tiles
+        tiles.forEach { tile ->
+            byTile.remove(tile)?.let { byKey.keys.removeAll(it) }
+            byTile[tile] = hashSetOf()
+        }
+        if (tiles.size == 1) {
+            // shortcut: replace old tile
+            byTile[tiles.single()] = entries.unzip().first.toHashSet()
+            byKey.putAll(entries)
+            return emptyList()
+        } else {
+            // put only what is inside tiles
+            // and return what wasn't put
+            val ignoredKeys = mutableListOf<T>() // could also use sth like filter or mapNotNull instead of mutable list
+            val bboxByTile = tiles.associateWith { it.asBoundingBox(tileZoom) }
+            entries.forEach { (key, pos) ->
+                for (tile in tiles) {
+                    if (bboxByTile[tile]!!.contains(pos)) {
+                        byTile[tile]!!.add(key)
+                        byKey[key] = pos
+                        return@forEach
+                    }
+                }
+                ignoredKeys.add(key)
+            }
+            return ignoredKeys
+        }
+    }
+
+    // bbox must fit in cache!
+    fun get(bbox: BoundingBox): List<T> {
+        val requiredTiles = bbox.asListOfEnclosingTilePos()
+        if (requiredTiles.size > maxTiles)
+            throw(IllegalArgumentException("trying to get more tiles than fit in cache"))
```

Comment:
I do not think this is a nice behavior. I wouldn't the app to crash in a case like that. It should just work as normal and then immediately trim some tiles (or don't put them in the first place, but IMO this is unnecessary code complexity+preliminary optimization for a case that **should** never happen).
Maybe log a info or warning, but that should be it.

Reply:
Ok, I changed it to increase in size return data normally, without the call to `trim()`.
Thus `SpatialCache.trim()` needs to be called after `SpatialCache.get()` **and** getting data from other caches.

Reply:
hmm

Reply:
> hmm

I guess this means you don't like having to manage cache size from outside the cache.
I also don't like it, maybe you can find a better solution here...